anymail/backends/mailpace.py

```diff
@@ -0,0 +1,190 @@
+from ..exceptions import AnymailRequestsAPIError
+from ..message import AnymailRecipientStatus
+from ..utils import CaseInsensitiveCasePreservingDict, get_anymail_setting
+from .base_requests import AnymailRequestsBackend, RequestsPayload
+
+
+class EmailBackend(AnymailRequestsBackend):
+    """
+    MailPace API Email Backend
+    """
+
+    esp_name = "MailPace"
+
+    def __init__(self, **kwargs):
+        """Init options from Django settings"""
+        esp_name = self.esp_name
+        self.server_token = get_anymail_setting(
+            "server_token", esp_name=esp_name, kwargs=kwargs, allow_bare=True
+        )
+        api_url = get_anymail_setting(
+            "api_url",
+            esp_name=esp_name,
+            kwargs=kwargs,
+            default="https://app.mailpace.com/api/v1/send",
+        )
+        if not api_url.endswith("/"):
+            api_url += "/"
+        super().__init__(api_url, **kwargs)
+
+    def build_message_payload(self, message, defaults):
+        return MailPacePayload(message, defaults, self)
+
+    def raise_for_status(self, response, payload, message):
+        # We need to handle 400 responses in parse_recipient_status
+        if response.status_code != 400:
+            super().raise_for_status(response, payload, message)
+
+    def parse_recipient_status(self, response, payload, message):
+        # Prepare the dict by setting everything to queued without a message id
+        unknown_status = AnymailRecipientStatus(message_id=None, status="queued")
+        recipient_status = CaseInsensitiveCasePreservingDict(
+            {recip.addr_spec: unknown_status for recip in payload.to_cc_and_bcc_emails}
+        )
+
+        parsed_response = self.deserialize_json_response(response, payload, message)
+
+        status_code = str(response.status_code)
+        json_response = response.json()
+
+        # Set the status_msg and id based on the status_code
+        if status_code == "200":
+            try:
+                status_msg = parsed_response["status"]
+                id = parsed_response["id"]
+            except (KeyError, TypeError) as err:
+                raise AnymailRequestsAPIError(
+                    "Invalid MailPace API response format",
+                    email_message=None,
+                    payload=payload,
+                    response=response,
+                    backend=self,
+                ) from err
+        elif status_code.startswith("4"):
+            status_msg = "error"
+            id = None
+
+        if status_msg == "queued":
+            # Add the message_id to all of the recipients
+            for recip in payload.to_cc_and_bcc_emails:
+                recipient_status[recip.addr_spec] = AnymailRecipientStatus(
+                    message_id=id, status="queued"
+                )
+        elif status_msg == "error":
+            if "errors" in json_response:
+                for field in ["to", "cc", "bcc"]:
```

Comment:
I'm a little confused by the logic in this section.

The goal is to provide a specific send status for each recipient, if known. 

I think this code will end up setting the sent status for *every* recipient based on the last reported error in the response (for *any* recipient). So if all "to" addresses were valid, but one of the cc's was on the blocked list, all of the "to" and the "cc" recipients would get reported as "blocked".

(Are there any cases where MailPace would send the message to some recipients, but not to others? Or if there's a problem with *any* to/cc/bcc recipient, is the entire message always rejected?)

Reply:
It's the latter - if there's a problem with any recipient (e.g. blocked or invalid) the email will be rejected (error http status from the endpoint) and not sent, as the entire email is technically not valid in that case.

Reply:
OK, thanks, that helps. I also did a little testing. [Sorry, this comment is kind of long—there are a bunch of cases to consider.]

The goal of `parse_recipient_status` is: based on the API response, for *each* individual "to" recipient, provide the send status *if we know it*, or "failed" if the message definitely wasn't sent to (queued for) that recipient but we can't exactly say why, or "unknown" if we're not sure whether or not the message was sent to that particular recipient. (This is especially important with some other ESPs that don't fail the entire message, but instead send to some recipients but not others.)

If that status information is also available for "cc" and "bcc" recipients, it's great to provide it. Otherwise, it's OK to just omit cc and bcc recipients.

In my testing, it looks like MailPace's API response currently reports:
* `{"errors":{"to":["is invalid"]}}` if at least one (but not necessarily all) of the "to" recipients is invalid
* `{"errors":{"to":["contains a blocked address"]}}` if at least one (but not necessarily all) of the "to" recipients is on the blocked list
* `{"errors":{"to":["contains a blocked address", "is invalid"]}}` if both conditions apply
* With multiple "to" addresses, there is no way to tell from the API response which of the addresses did or didn't cause these errors
* It seems like errors in "cc" or "bcc" addresses are not currently reported—the message is queued and delivered to the "to" recipients. [Aside: it looks like delivery is attempted for a blocked cc. An invalid cc field seems to cause the message to get stuck in the Queued status. I'm guessing these might be bugs?]

With that in mind, here are some scenarios `parse_recipient_status` needs to cover:

* One or more "to" recipients, no errors
  * Current implementation: all "to" recipients are reported "queued", which is the desired result
* Exactly one "to" recipient, invalid address
  * Current implementation: the recipient is reported as "invalid", which is the desired result
* Exactly one "to" recipient, on the blocked list 
  * Current implementation: the recipient is reported as "rejected", which is the desired result
* Multiple "to" recipients, some invalid and/or blocked (and some not)
  * Current implementation: *all* "to" recipients are reported as either "invalid" or "rejected", even if that individual recipient was valid. This is misleading. (Also, when both errors apply, whichever happens to be last in the errors list determines the status for all recipients.)
  * Desired result: with multiple recipients and some invalid/blocked errors, use status "failed" for all "to" recipients. (We know the message wasn't sent to that recipient, but for each individual address we can't be sure why—or even that there was a problem with *that* address.)
* Any of the above, but also with "cc" and "bcc" recipients
  * Current implementation: "cc" and "bcc" recipients are given the same status as the "to" recipient(s) ("queued" or "invalid" or "rejected"), which may be misleading (because MailPace doesn't currently report errors on cc and bcc, and the cc/bcc recipients aren't necessarily invalid/blocked just because one of the to recipients is)
  * Desired result: probably easiest to just omit "cc" and "bcc" recipients. (Could instead give them status "unknown" if the "to" is "queued", or "failed" if the "to" is invalid/blocked, but that doesn't add much value.)
* Zero "to" recipients, one or more "cc" and/or "bcc" recipients
  * [Developers sometimes try to use cc/bcc-only messages to implement something like a mailing list]
  * Current implementation: all "cc" and "bcc" recipients are given status "invalid", which will be misleading (the cc and bcc recipients are likely valid email addresses)
  * Desired result: since MailPace doesn't support messages without "to", just let its `{"errors":{"to":["undefined field"]}}` response be reported as an AnymailRequestsAPIError. (Don't try to handle this through recipient status.)
* More than 50 "to" recipients
  * Current implementation: all recipients are reported as "invalid", which is misleading
  * Desired result: just let MailPace's `"number of email addresses exceeds maximum volume"` response be reported as an AnymailRequestsAPIError. (Don't try to handle this through recipient status.)

Here's a suggested approach that I think achieves all of this:
* Replace `to_cc_and_bcc_emails` with just `to_emails`
* If `response.status_code == 200`, return `status = "queued"` for all `to_emails`
  * Is it necessary to check `parsed_response["status"] == "queued"`? (Could there ever be any other status with a 200 response?)
* If `response.status_code == 400` and the *only* field in `parsed_response["errors"]` is `"to"` and the errors in `parsed_response["errors"]["to"]` are *only* `"is invalid"` and/or `"contains a blocked address"`:
  * If there is exactly one `to_emails`:
    * If `"is invalid"`, return `status = "invalid"` for all `to_emails`
    * If `"contains a blocked address"`, return `status = "rejected"` for all `to_emails`
    * If both, return `status = "failed"` for all `to_emails`
  * Else (more than one `to_emails`) return `status = "failed"` for all `to_emails` (we don't know which are invalid or rejected)
* In *all other* cases, raise an `AnymailRequestsAPIError` for the response
  * This includes all other problems that might be reported in `parsed_response["errors"]["to"]`
  * This also covers the case where MailPace later starts reporting errors in cc/bcc fields
* (Wrap all `parsed_response[...]` access to convert KeyError and TypeError to `AnymailRequestsAPIError("Unexpected MailPace API response format"...)`)

Does that make sense? Have I missed anything?

Reply:
Thanks @medmunds  - my suggestion here is for us to fix the issues with non-reporting of errors and blocked emails to cc and bcc'd recipients first (plus the other bug reported below), then come back and fix the PR appropriately. We'll get back shortly on that.

Reply:
> fix the issues with non-reporting of errors and blocked emails to cc and bcc'd recipients first

Sounds good, thanks.